sphinx/domains/std.py

```diff
@@ -451,17 +460,38 @@ def run(self) -> List[Node]:
             subnode = addnodes.production(rule)
             subnode['tokenname'] = name.strip()
             if subnode['tokenname']:
-                idname = nodes.make_id('grammar-token-%s' % subnode['tokenname'])
+                idname = 'grammar-token2-%s_%s' \
+                         % (nodes.make_id(productionGroup), nodes.make_id(name))
                 if idname not in self.state.document.ids:
                     subnode['ids'].append(idname)
+
+                idnameOld = nodes.make_id('grammar-token-' + name)
+                if idnameOld not in self.state.document.ids:
+                    subnode['ids'].append(idnameOld)
```

Comment:
Is this old node_id still needed?

Reply:
That depends. If someone has copied a link to a rule, then it would be broken without this code. However, there are no explicit permalinks in production lists, link there is on the various domain directives. So I guess that technically there is no promise that such links will continue to work :-).

Reply:
Understood. Indeed, an URL of old `:token:` role causes an error. +0 for keeping this. My large concern is that codes for compatibility makes our code complex. So it would be better to remove this in future. I've sometime confused by "historical" code...

Reply:
Indeed, some cleanup would be nice. Though I'm not sure what the best approach is so I'll let the old ID be there for now.